File: 34.find-first-and-last-position-of-element-in-sorted-array.js
/*
 * @lc app=leetcode id=34 lang=javascript
 *
 * [34] Find First and Last Position of Element in Sorted Array
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var searchRange = function (nums, target) {
  const searchLowerBound = (nums, target) => {
    let i = 0;
    let j = nums.length - 1;
    let res = -1;
    while (i <= j) {
      const m = i + ((j - i) >> 1);
      if (nums[m] < target) i = m + 1;
      else if (nums[m] > target) j = m - 1;
      else {
        res = m;
        j = m - 1;
      }
    }
    return res;
  };

  const searchUpperBound = (nums, target) => {
    let i = 0;
    let j = nums.length - 1;
    let res = -1;
    while (i <= j) {
      const m = i + ((j - i) >> 1);
      if (nums[m] < target) i = m + 1;
      else if (nums[m] > target) j = m - 1;
      else {
        res = m;
        i = m + 1;
      }
    }
    return res;
  };

  return [searchLowerBound(nums, target), searchUpperBound(nums, target)];
};
// @lc code=end

File: solutions/34.find-first-and-last-position-of-element-in-sorted-array.rb
#
# @lc app=leetcode id=34 lang=ruby
#
# [34] Find First and Last Position of Element in Sorted Array
#

# @lc code=start
# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer[]}
def search_range(nums, target)
  lower_bound = begin
    i = 0
    j = nums.length - 1
    while i <= j
      m = i + (j - i) / 2
      if nums[m] < target
        i = m + 1
      else
        j = m - 1
        found = m if nums[m] == target
      end
    end
    found || -1
  end

  return [-1, -1] if lower_bound == -1 # cannot early return in begin ...end block

  higher_bound = begin
    i = lower_bound
    j = nums.length - 1
    while i <= j
      m = i + (j - i) / 2
      if nums[m] <= target
        i = m + 1
        found = m if nums[m] == target
      else
        j = m - 1
      end
    end
    found
  end
  [lower_bound, higher_bound]
end
# @lc code=end

search_range([5,7,7,8,8,10], 8)
